R/read_delim.R

```diff
@@ -420,7 +432,9 @@ read_delimited <- function(file, tokenizer, col_names = TRUE, col_types = NULL,
 
   ds <- datasource(data, skip = spec$skip, skip_empty_rows = skip_empty_rows, comment = comment, skip_quote = skip_quote)
 
-  if (is.null(col_types) && !inherits(ds, "source_string") && !is_testing()) {
+  has_col_types <- !is.null(col_types)
+
+  if (((is.null(show_col_types) && !has_col_types) || isTRUE(show_col_types)) && !inherits(ds, "source_string") && !is_testing()) {
```

Comment:
This logic is a bit confusing, basically if `show_col_types` is `NULL`, which is the default then we only want to show the column types if they are not explicitly set, or if we are running the tests or getting data from a string.

Suggestions welcome on ways to make it more understandable.

Reply:
It seems like the presence of `is_testing()` will be an obstacle to snapshot testing, which would be nice for the col spec stuff, so we might want to refactor a bit. If I end up adding tests, I could work on that.